src/validation.cpp

```diff
@@ -1393,56 +1417,93 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
+                    // Erase any previous invalid results for this transaction. For example, this
+                    // could be a low-feerate tx that has just been bumped.
+                    individual_results_nonfinal.erase(subpackage_wtxid);
+                    packageified.Exclude(subpackage_tx);
+                }
+            }
+            // Note that it's possible for transactions to have been submitted to the mempool even
+            // if subpackage_result.m_state.IsInvalid(). If IsValid(), fine to move on.
+            if (subpackage_result.m_state.IsValid()) continue;
+
+            // Another "final" validation result is if the tx failed for a non-policy reason.
+            const auto single_res_it = subpackage_result.m_tx_results.find(wtxid);
+            if (single_res_it != subpackage_result.m_tx_results.end()) {
+                const auto single_res = single_res_it->second;
+                if (single_res.m_state.GetResult() != TxValidationResult::TX_MEMPOOL_POLICY &&
+                    single_res.m_state.GetResult() != TxValidationResult::TX_MISSING_INPUTS) {
+                    // Package validation policy only differs from individual policy in its evaluation
+                    // of feerate. For example, if a transaction fails here due to violation of a
+                    // consensus rule, the result will not change when it is submitted as part of a
+                    // package. To minimize the amount of repeated work, unless the transaction fails
+                    // due to feerate or missing inputs (its parent is a previous transaction in the
+                    // package that failed due to feerate), don't run package validation. Note that this
+                    // decision might not make sense if different types of packages are allowed in the
+                    // future.  Continue individually validating the rest of the transactions, because
+                    // some of them may still be valid.
+                    quit_early = true;
+                    package_state_quit_early.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
+                    individual_results_nonfinal.emplace(wtxid, single_res);
+                    packageified.Ban(tx);
+                }
                 individual_results_nonfinal.emplace(wtxid, single_res);
-                txns_package_eval.push_back(tx);
             }
         }
     }
 
     // Quit early because package validation won't change the result or the entire package has
-    // already been submitted.
-    if (quit_early || txns_package_eval.empty()) {
+    // already been submitted. Since this is an ancestor package, if the child is in, that means all
+    // the other transactions in the package are as well. We check for the child by txid because
+    // same-txid-different-witness is an acceptable case for deduplication in the loop above.
+    if (quit_early || m_pool.exists(GenTxid::Txid(child->GetHash()))) {
```

Comment:
If we change `GetAncestorSet` to an optional and store banned-ness, then we can use it to check if there is nothing left to be submitted directly by calling `txns_package_eval.empty()`.

```suggestion
    if (packageified.HasBan() || txns_package_eval.empty()) {
```

Or, if we keep GetAncestorSet as is, just do `txns_package_eval.empty()` only? It will be empty in both the cases where it has a banned input, or where all transactions have been excluded?

Then we can just get rid of quit_early?

Reply:
Good idea, got rid of quit_early and exists(). Using txns_package_eval std::nullopt or empty